src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,240 @@
+default_initial_message(inds_e) = ITensor[denseblocks(delta(inds_e))]
+default_messages(ptn::PartitionedGraph) = Dictionary()
+function default_message_update(contract_list::Vector{ITensor}; kwargs...)
+  return contract_exact(contract_list; kwargs...)
+end
+default_message_update_kwargs() = (; normalize=true, contraction_sequence_alg="optimal")
+@traitfn default_bp_maxiters(g::::(!IsDirected)) = is_tree(g) ? 1 : nothing
+@traitfn function default_bp_maxiters(g::::IsDirected)
+  return default_bp_maxiters(undirected_graph(underlying_graph(g)))
+end
+function message_diff(message_a::Vector{ITensor}, message_b::Vector{ITensor})
+  lhs, rhs = contract(message_a), contract(message_b)
+  return 0.5 *
+         norm((denseblocks(lhs) / sum(diag(lhs))) - (denseblocks(rhs) / sum(diag(rhs))))
+end
+
+struct BeliefPropagationCache{PTN,MTS,DM}
+  partitioned_itensornetwork::PTN
+  messages::MTS
+  default_message::DM
+end
+
+#Constructors...
+function BeliefPropagationCache(
+  ptn::PartitionedGraph;
+  messages=default_messages(ptn),
+  default_message=default_initial_message,
+)
+  return BeliefPropagationCache(ptn, messages, default_message)
+end
+
+function BeliefPropagationCache(tn::ITensorNetwork, partitioned_vertices; kwargs...)
+  ptn = PartitionedGraph(tn, partitioned_vertices)
+  return BeliefPropagationCache(ptn; kwargs...)
+end
+
+function partitioned_itensornetwork(bp_cache::BeliefPropagationCache)
+  return bp_cache.partitioned_itensornetwork
+end
+messages(bp_cache::BeliefPropagationCache) = bp_cache.messages
+default_message(bp_cache) = bp_cache.default_message
+function tensornetwork(bp_cache::BeliefPropagationCache)
+  return unpartitioned_graph(partitioned_itensornetwork(bp_cache))
+end
+
+#Forward from partitioned graph
+for f in [
+  :(NamedGraphs.partitioned_graph),
+  :(NamedGraphs.partitionedge),
+  :(NamedGraphs.partitionvertices),
+  :(NamedGraphs.vertices),
+  :(NamedGraphs.boundary_partitionedges),
+]
+  @eval begin
+    function $f(bp_cache::BeliefPropagationCache, args...; kwargs...)
+      return $f(partitioned_itensornetwork(bp_cache), args...; kwargs...)
+    end
+  end
+end
+
+function linkinds(bp_cache::BeliefPropagationCache, edge::PartitionEdge)
+  ptn = partitioned_itensornetwork(bp_cache)
+  src_e_itn = subgraph(ptn, src(edge))
+  dst_e_itn = subgraph(ptn, dst(edge))
+  return commoninds(src_e_itn, dst_e_itn)
+end
```

Comment:
Is the plan to define `linkinds(partioned_tn::PartitionedGraph, edge::PartitionEdge)` and have this call that?

Reply:
Yes, just adding that now!